src/commands/register.ts

```diff
@@ -2,142 +2,161 @@ import { flags } from '@oclif/command';
+    await this.post_component_to_api(component_dto, selected_account.id);
+    cli.action.stop(chalk.green(`Successfully registered component`));
+  }
+
+
+  private async build_docker_image_if_necessary(config_path: string, service_config: ServiceConfig, image_tag: string) {
+    // if the image field is set, we just take their image as is
+    if (service_config.getImage()) {
+      return service_config.getImage();
+    }
+
+    // otherwise we build and push the image to our repository
+    const image = await this.build_image(config_path, service_config, image_tag);
+    await this.push_image(image);
+    const digest = await this.get_digest(image);
+
+    // we don't need the tag on our image because we use the digest as the key
+    const image_without_tag = strip_tag_from_image(image);
+    return `${image_without_tag}@${digest}`;
+  }
+
+
+  private async build_image(config_path: string, service_config: ServiceConfig, image_tag: string) {
+    const build_context = service_config.getBuild().context;
+    if (!build_context) {
+      throw new Error(`Service ${service_config.getName()} does not specify an image or a build.context. It must contain one or the other.`);
+    }
+    try {
+      const component_path = fs.lstatSync(config_path).isFile() ? path.dirname(config_path) : config_path;
+      const build_path = path.resolve(component_path, build_context);
+
+      return await buildImage(build_path, image_tag);
+    } catch (err) {
+      cli.action.stop(chalk.red(`Build failed`));
+      this.log(`Docker build failed. If an image is not specified in your service config or as a flag, then a Dockerfile must be present`);
+      throw new Error(err);
     }
   }
 
-  private async post_service_to_api(dto: CreateServiceVersionInput, account_id: string): Promise<any> {
-    const { data: service_digest } = await this.app.api.post(`/accounts/${account_id}/services`, dto);
-    return service_digest;
+
+  private async push_image(image: string) {
+    cli.action.start(chalk.blue(`Pushing Docker image for ${image}`));
+    try {
+      await pushImage(image);
+    } catch (err) {
+      cli.action.stop(chalk.red(`Push failed for image ${image}`));
+      throw new Error(err);
+    }
+    cli.action.stop(chalk.green(`Successfully pushed Docker image for ${image}`));
+  }
+
+
+  private async get_digest(image: string) {
+    cli.action.start(chalk.blue(`Running \`docker inspect\` on the given image: ${image}`));
+    const digest = await getDigest(image).catch(e => {
+      cli.action.stop(chalk.red(`Inspect failed`));
+      throw new Error(`The image is not reachable by docker: ${image}`);
+    });
+    cli.action.stop(chalk.green(`Image verified`));
+  }
+
+
+  private async post_component_to_api(dto: PutComponentVersionDto, account_id: string): Promise<any> {
+    const { data: component_version } = await this.app.api.post(`/accounts/${account_id}/components`, dto);
```

Comment:
Can this just go on the line where it's used above? It also doesn't look like anything really needs to be returned